### PR TITLE
MM-47064 :  Migrate "components/suggestion/suggestion.jsx" to Typescript

### PR DESCRIPTION
**Pull request by AbhinavVihan**
#### Summary
Suggestion.jsx to typescript migration

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/21099
Fixes https://mattermost.atlassian.net/browse/MM-47064

#### Related Pull Requests
None

#### Screenshots
NA

#### Release Note
None
```release-note
Suggestion.jsx to typescript migration
```

### CODE REVIEW

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx [hunk 1]
```diff
@@ -16,18 +16,19 @@ import 'moment';
...
-    locale: string;
+    locale?: string;
+    currentDate?: any;
 }
 
 export default class SearchDateSuggestion extends Suggestion {
     handleDayClick = (day: Date) => {
         const dayString = day.toISOString().split('T')[0];
-        this.props.onClick(dayString, this.props.matchedPretext);
+        this.props.onClick!(dayString, this.props.matchedPretext);
```
Comment (M-ZubairAhmed) [inline components/suggestion/search_date_suggestion/search_date_suggestion.tsx:26]: It more readable to just check for undefined
```suggestion
        if (this.props.onClick) {
            this.props.onClick(dayString, this.props.matchedPretext);
        }
```

File: components/suggestion/suggestion.tsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+export default class Suggestion extends React.PureComponent<Props> {
+    static baseProps = {
+        role: 'button',
+        tabIndex: -1,
+    };
+
+    handleClick = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+        e.preventDefault();
+
+        this.props.onClick!(this.props.term, this.props.matchedPretext);
```
Comment (M-ZubairAhmed) [inline components/suggestion/suggestion.tsx:28]: https://github.com/mattermost/mattermost-webapp/pull/11173#discussion_r981103665

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/suggestion/suggestion.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    };
+
+    handleMouseMove = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
+        e.preventDefault();
+
+        this.props.onMouseMove!(this.props.term);
```
Comment (devinbinnie) [inline components/suggestion/suggestion.tsx:34]: Let's do the same as Zubair's comment here as well (wrt. optional chaining)

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx (hunk 1 above)
Comment (devinbinnie) [reply components/suggestion/search_date_suggestion/search_date_suggestion.tsx:26]: Actually we can do this even better with a coalesce operator
```suggestion
        this.props.onClick?.(dayString, this.props.matchedPretext);
```

File: components/suggestion/suggestion.tsx (hunk 3 above)
Comment (devinbinnie) [reply components/suggestion/suggestion.tsx:34]:
```suggestion
        this.props.onMouseMove?.(this.props.term);
```

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx [hunk 4]
```diff
@@ -55,7 +56,7 @@ export default class SearchDateSuggestion extends Suggestion {
             };
         }
 
-        const locale = this.props.locale.toLowerCase();
+        const locale = this.props.locale && this.props.locale.toLowerCase();
```
Comment (devinbinnie) [inline components/suggestion/search_date_suggestion/search_date_suggestion.tsx:59]:
```suggestion
        const locale = this.props.locale?.toLowerCase();
```

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx [hunk 5]
```diff
@@ -16,18 +16,19 @@ import 'moment';
...
     handleDayClick = (day: Date) => {
         const dayString = day.toISOString().split('T')[0];
-        this.props.onClick(dayString, this.props.matchedPretext);
+        this.props.onClick?.(dayString, this.props.matchedPretext);
     }
 
     componentDidMount() {
         //the naming scheme of momentjs packages are all lowercases
-        const locale = this.props.locale.toLowerCase();
+        const locale = this.props.locale && this.props.locale.toLowerCase();
```
Comment (devinbinnie) [inline components/suggestion/search_date_suggestion/search_date_suggestion.tsx:31]: We can do this one using the coalesce as well.
```suggestion
        const locale = this.props.locale?.toLowerCase();
```

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx [hunk 6]
```diff
@@ -16,18 +16,19 @@ import 'moment';
 const loadedLocales: Record<string, moment.Locale> = {};
 
 type Props = {
-    locale: string;
+    locale?: string;
+    currentDate?: any;
```
Comment (M-ZubairAhmed) [inline components/suggestion/search_date_suggestion/search_date_suggestion.tsx:20]: can we get a better type here?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/suggestion/suggestion.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import {UserProfile} from './command_provider/app_command_parser/app_command_parser_dependencies';
+
+type Props = {
+    onClick?: (term: string, matchedPretext: string) => boolean;
+    onMouseMove?: (term: string) => void;
+    term: string;
+    matchedPretext: string;
+    isSelection?: boolean;
+    item: any;
```
Comment (AbhinavVihan) [inline components/suggestion/suggestion.tsx:15]: what should be a better type here.
Item is being used in many provider files.
Its type is different in almost each file.

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx [hunk 8]
```diff
@@ -3,11 +3,11 @@
 
 import React from 'react';
 
-import DayPicker from 'react-day-picker';
+import DayPicker, {Modifier} from 'react-day-picker';
```
Comment (M-ZubairAhmed) [inline components/suggestion/search_date_suggestion/search_date_suggestion.tsx:6]: should this be `DayModifiers` ?

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: components/suggestion/search_channel_suggestion/search_channel_suggestion.tsx [hunk 9]
```diff
@@ -68,14 +68,15 @@ function itemToName(item: Channel, currentUser: string): {icon: React.ReactEleme
...
+        const {isSelection, teammate, currentUser} = this.props;
+        const item = this.props.item as Channel;
 
         let className = 'suggestion-list__item';
         if (isSelection) {
             className += ' suggestion--selected';
         }
 
-        const nameObject = itemToName(item, currentUser);
+        const nameObject = itemToName(item, currentUser as string);
```
Comment (devinbinnie) [inline components/suggestion/search_channel_suggestion/search_channel_suggestion.tsx:79]: Nit: Just curious why we need the cast for `currentUser`? Does it not think it's a `string`?

**Review by devinbinnie — APPROVED** [review]
Thanks @AbhinavVihan, LGTM other than Zubair's comments. I had one non-blocking, but nothing major.

File: components/suggestion/search_channel_suggestion/search_channel_suggestion.tsx (hunk 9 above)
Comment (AbhinavVihan) [reply components/suggestion/search_channel_suggestion/search_channel_suggestion.tsx:79]: No, it is optional in suggestion.tsx. (so its type here is string | undefined);
should this be written in an if statement instead?

File: components/suggestion/search_date_suggestion/search_date_suggestion.tsx (hunk 8 above)
Comment (AbhinavVihan) [reply components/suggestion/search_date_suggestion/search_date_suggestion.tsx:6]: @M-ZubairAhmed, the type of modifier is
Screenshot from 2022-10-14 10-01-40
So it gives an error **Type '{ today: DayModifiers; } | undefined' is not assignable to type 'Partial<Modifiers> | undefined'.**

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]
Hey can you please fix the conflicts 😅

File: components/suggestion/search_channel_suggestion/search_channel_suggestion.tsx (hunk 9 above)
Comment (devinbinnie) [reply components/suggestion/search_channel_suggestion/search_channel_suggestion.tsx:79]: You can instead use a coalesce operator to pass in a blank string instead. Assuming that the method actually cares that it's an actual string and not a null.
```suggestion
        const nameObject = itemToName(item, currentUser ?? '');
```

**Comment by M-ZubairAhmed** [comment]
@AbhinavVihan there seems to be a lot of conflicts. If you are still planning to work on this i would recommend opening another a fresh pr from latest master. It would make things a lot easier for you. 😄

**Comment by sinansonmez** [comment]
> @AbhinavVihan there seems to be a lot of conflicts. If you are still planning to work on this i would recommend opening another a fresh pr from latest master. It would make things a lot easier for you. 😄

@AbhinavVihan please let us know how would you like to proceed on this PR

**Comment by AbhinavVihan** [comment]
I would like to close this PR and create a fresh new one please

**Comment by M-ZubairAhmed** [comment]
feel free to do that @AbhinavVihan

**Pull request closed without merging**